pennylane/transforms/decompose.py

```diff
@@ -0,0 +1,121 @@
+    """Decomposes operations found in a quantum circuit into a desired gate set.
+
+    Args:
+        tape (QuantumScript or QNode or Callable): a quantum circuit.
+        gate_set (Callable or set, optional): A set of decomposition gates or a condition they satisfy. Defaults to None.
+        max_expansion (int, optional): The maximum depth of the expansion. Defaults to None.
+
+    Returns:
+        qnode (QNode) or quantum function (Callable) or tuple[List[QuantumScript], function]:
+
+        The decomposed circuit. The output type is explained in :func:`qml.transform <pennylane.transform>`.
+
+    Raises:
+        Exception: Type defaults to ``qml.DeviceError``.
+            Raised if an operator is not accepted and does not define a decomposition, or if
+            the decomposition enters an infinite loop and raises a ``RecursionError``.
+
+    .. seealso:: :func:`~.pennylane.devices.preprocess.decompose` for a transform that performs the same job but is designed particularly for specific device architectures.
+
+    **Examples:**
+
+    >>> tape = qml.tape.QuantumScript([qml.IsingXX(1.2, wires=(0,1))], [qml.expval(qml.Z(0))])
+    >>> batch, fn = decompose(tape, gate_set = {"CNOT", "RX", "RZ"})
+    >>> batch[0].circuit
+    [CNOT(wires=[0, 1]),
+    RX(1.2, wires=[0]),
+    CNOT(wires=[0, 1]),
+    expval(Z(0))]
+
+    >>> @partial(decompose, gate_set = lambda obj: len(obj.wires) <= 2)
+    >>> @qml.qnode(device)
+    >>> def circuit():
+    >>>     qml.Toffoli(wires = range(2))
+    >>>
+    >>> print(qml.draw(circuit)())
+    0: ───────────╭●───────────╭●────╭●──T──╭●─┤
+    1: ────╭●─────│─────╭●─────│───T─╰X──T†─╰X─┤
+    2: ──H─╰X──T†─╰X──T─╰X──T†─╰X──T──H────────┤
+
+    """
+    if gate_set is None:
+        gate_set = set(qml.ops.__all__)
+
+    if isinstance(gate_set, (list, tuple)):
+        gate_set = set(gate_set)
+
+    def decomposer(op):
+        return op.decomposition()
+
+    def stopping_condition(op):
+        if not isinstance(op, qml.operation.Operator):
+            return True
+        if not op.has_decomposition:
+            return True
+        if isinstance(gate_set, set):
+            return op.name in gate_set
+        return gate_set(op)
+
+    if tape.operations and isinstance(tape[0], StatePrepBase):
+        prep_op = [tape[0]]
+    else:
+        prep_op = []
+
+    if all(stopping_condition(op) for op in tape.operations[len(prep_op) :]):
+        return (tape,), null_postprocessing
+
+    try:
+        new_ops = [
+            final_op
+            for op in tape.operations[len(prep_op) :]
+            for final_op in qml.devices.preprocess._operator_decomposition_gen(
+                op,
+                stopping_condition,
+                decomposer=decomposer,
+                max_expansion=max_expansion,
+                error=qml.operation.DecompositionUndefinedError,
+            )
+        ]
+    except RecursionError as e:
+        raise qml.DeviceError(
```

Comment:
Since this transform isn't meant to be used in the context of a device, a `RecursionError` probably makes more sense. Here we could suggest setting `max_expansion` to stop the decomposition after a set number of attempts as a solution to the RecursionError (bad solution on a device, but I think in the spirit of the intended behaviour for this transform).